src/kernel/bitcoinkernel.cpp

```diff
@@ -1,10 +1,1217 @@
+
+    auto byte_array{new kernel_ByteArray{
+        .data = new unsigned char[script_pubkey->size()],
+        .size = script_pubkey->size(),
+    }};
+
+    std::memcpy(byte_array->data, script_pubkey->data(), byte_array->size);
+    return byte_array;
+}
+
+void kernel_script_pubkey_destroy(kernel_ScriptPubkey* script_pubkey)
+{
+    if (script_pubkey) {
+        delete cast_script_pubkey(script_pubkey);
+    }
+}
+
+kernel_TransactionOutput* kernel_transaction_output_create(const kernel_ScriptPubkey* script_pubkey_, int64_t amount)
+{
+    const auto& script_pubkey{*cast_script_pubkey(script_pubkey_)};
+    const CAmount& value{amount};
+    auto tx_out{new CTxOut(value, script_pubkey)};
+    return reinterpret_cast<kernel_TransactionOutput*>(tx_out);
+}
+
+void kernel_transaction_output_destroy(kernel_TransactionOutput* output)
+{
+    if (output) {
+        delete cast_transaction_output(output);
+    }
+}
+
+bool kernel_verify_script(const kernel_ScriptPubkey* script_pubkey_,
+                         const int64_t amount_,
+                         const kernel_Transaction* tx_to,
+                         const kernel_TransactionOutput** spent_outputs_, size_t spent_outputs_len,
+                         const unsigned int input_index,
+                         const unsigned int flags,
+                         kernel_ScriptVerifyStatus* status)
+{
+    const CAmount amount{amount_};
+    const auto& script_pubkey{*cast_script_pubkey(script_pubkey_)};
+
+    if (!verify_flags(flags)) {
+        if (status) *status = kernel_SCRIPT_VERIFY_ERROR_INVALID_FLAGS;
+        return false;
+    }
+
+    if (!is_valid_flag_combination(flags)) {
+        if (status) *status = kernel_SCRIPT_VERIFY_ERROR_INVALID_FLAGS_COMBINATION;
+        return false;
+    }
+
+    if (flags & kernel_SCRIPT_FLAGS_VERIFY_TAPROOT && spent_outputs_ == nullptr) {
+        if (status) *status = kernel_SCRIPT_VERIFY_ERROR_SPENT_OUTPUTS_REQUIRED;
+        return false;
+    }
+
+    const CTransaction& tx{*cast_transaction(tx_to)};
+    std::vector<CTxOut> spent_outputs;
+    if (spent_outputs_ != nullptr) {
+        if (spent_outputs_len != tx.vin.size()) {
+            if (status) *status = kernel_SCRIPT_VERIFY_ERROR_SPENT_OUTPUTS_MISMATCH;
+            return false;
+        }
+        spent_outputs.reserve(spent_outputs_len);
+        for (size_t i = 0; i < spent_outputs_len; i++) {
+            const CTxOut& tx_out{*reinterpret_cast<const CTxOut*>(spent_outputs_[i])};
+            spent_outputs.push_back(tx_out);
+        }
+    }
+
+    if (input_index >= tx.vin.size()) {
+        if (status) *status = kernel_SCRIPT_VERIFY_ERROR_TX_INPUT_INDEX;
+        return false;
+    }
+    PrecomputedTransactionData txdata{tx};
+
+    if (spent_outputs_ != nullptr && flags & kernel_SCRIPT_FLAGS_VERIFY_TAPROOT) {
+        txdata.Init(tx, std::move(spent_outputs));
+    }
+
+    return VerifyScript(tx.vin[input_index].scriptSig,
```

Comment:
We discussed during the last workshop that ideally we don't have any status codes here at all. But the problem is annoying to tackle. You'd probably want to pass this function a script verify object that has already passed through the required pre-checks. But then you have to either copy the objects into this object, or give ownership up to that object, which I don't think is desirable. The alternative to that is having a function with the same signature that you can call to check the arguments. But then you're forced to check them here again. I'm coming around to the option of replacing the status codes with log messages, but then we're sacrificing a bit of responsiveness to the developer.

Reply:
What do you think of doing something like this instead: https://github.com/TheCharlatan/bitcoin/commit/6323d7b072de5b13ab25aaa29e02332c44808b62

Reply:
> You'd probably want to pass this function a script verify object that has already passed through the required pre-checks.

I 100% agree with this approach. Adding extra types makes the API more cumbersome to use, but I think it does make it more safe, and the extra verbosity should be quite easy to hide in client libraries.

> But then you're forced to check them here again.

I think that can be avoided by having the prechecks function return a `ScriptPreChecksPassed*` (which references the object it verified, but doesn't copy it) and then requiring that as an extra argument (extra as a way to address lifetime issues) to the `kernel_verify_script` function? This doesn't prevent runtime issues (e.g. re-using `PreChecksPassed` pointers, which is easily verifiable at runtime) or segfaults, but at least it adds some compile-time checks to guide the user to using the API safely, and it can be done without any copies or changing ownership?

> What do you think of doing something like this instead: https://github.com/TheCharlatan/bitcoin/commit/6323d7b072de5b13ab25aaa29e02332c44808b62

I'm not convinced. I think not requiring the user to deal with status codes (or strings) is a good philosophy, but if it's optional anyway, then passing a bool to add a log entry feels like a much worse interface? And I think it's inferior to this approach:

> The alternative to that is having a function with the same signature that you can call to check the arguments.

Reply:
I'll propose a pre-check function and object later today then :)